### PR TITLE
FT2023-345: Addition of the my_field module.

### PR DESCRIPTION
**Pull request by Arijit-Acharya-innoraft**

### CODE REVIEW

File: web/modules/custom/my_field/config/schema/my_field.schema.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+field.field_color_field.value:
+  type: sequence
+  label: 'Default value'
+  sequence:
+    - type: mapping
```
Comment (vivekpan27) [inline web/modules/custom/my_field/config/schema/my_field.schema.yml]: Should there be a `-` here?

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BackgroundColorFormatter.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      $attributes = new Attribute();
+      $attributes->setAttribute('style', 'background-color: ' . $color);
+      $elements[$delta] = [
+        '#type' => 'html_tag',
+        '#tag' => 'div',
+        '#value' => $color,
+        '#attributes' => $attributes->toArray(),
+        // '#attributes'=>[
+          // 'style' => 'background-color: ' . $color . ';',
+        // ]
```
Comment (vivekpan27) [inline web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BackgroundColorFormatter.php]: Please remove commented code.

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BodyValueFormatter.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  public function viewElements(FieldItemListInterface $items, $langcode) {
+
+    $elements = [];
+    foreach ($items as $delta => $item) {
+      // Storing the color.
+      $color = $item->hexcode;
+      // Storing that particuar entity.
+      $entity = $item->getEntity();
+      // Fetching that entity's body value.
+      $body_value = $entity->body->value;
```
Comment (vivekpan27) [inline web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BodyValueFormatter.php]: What if $entity does not have body attribute?
Please use condition here.

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/TextColorFormatter.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+use Drupal\Core\Field\FieldItemListInterface;
+use Drupal\Core\Field\FormatterBase;
+use Drupal\Core\Template\Attribute;
+
+/**
+ * Plugin implementation of the 'text color' formatter.
+ *
+ * @FieldFormatter(
+ *   id = "text_formatter",
```
Comment (vivekpan27) [inline web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/TextColorFormatter.php]: text_color_formatter?

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/TextColorFormatter.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  public function viewElements(FieldItemListInterface $items, $langcode) {
+
+    $elements = [];
+    foreach ($items as $delta => $item) {
+      // Storing the color.
+      $color = $item->hexcode;
+      // Storing that particuar entity.
+      $entity = $item->getEntity();
+      // Fetching that entity's body value.
+      $body_value = $entity->body->value;
```
Comment (vivekpan27) [inline web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/TextColorFormatter.php]: Same comment as before.

File: web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ * )
+ */
+class FieldColor extends FieldItemBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function schema(FieldStorageDefinitionInterface $field_definition) {
+
+    // Columns contains thevalues that the field will store.
```
Comment (vivekpan27) [inline web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php]: the values

File: web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    return [
+      'columns' => [
```
Comment (vivekpan27) [inline web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php:30]: why `columns` key is added?

File: web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      ->setLabel('Hexcode Value');
+    return $property;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isEmpty() {
+    $hexcode = $this->get('hexcode')->getValue();
+    return $hexcode === NULL || $hexcode === '';
```
Comment (vivekpan27) [inline web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php]: `return empty($hexcode)`.

File: web/modules/custom/my_field/src/Plugin/Field/FieldWidget/ColorPickerWidget.php [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+ *     "field_color"
+ *   }
+ * )
+ */
+class ColorPickerWidget extends WidgetBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
```
Comment (vivekpan27) [inline web/modules/custom/my_field/src/Plugin/Field/FieldWidget/ColorPickerWidget.php:26]: Please add a screenshot.

File: web/modules/custom/my_field/src/Plugin/Field/FieldWidget/HexcodeWidget.php [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+ * )
+ */
+class HexcodeWidget extends WidgetBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
+
+    $value = $items[$delta]->hexcode ?? '';
```
Comment (vivekpan27) [inline web/modules/custom/my_field/src/Plugin/Field/FieldWidget/HexcodeWidget.php:28]: Please add screenshot of the widget

File: web/modules/custom/my_field/src/Plugin/Field/FieldWidget/RgbColorWidget.php [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+   * Massages an array of RGB color values into hexadecimal color codes.
+   *
+   * @param array $values
+   *   An array of color values, each containing 'red', 'green' and 'blue' keys.
+   * @param array $form
+   *   The form structure.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   */
+  public function massageFormValues(array $values, array $form, FormStateInterface $form_state) {
```
Comment (vivekpan27) [inline web/modules/custom/my_field/src/Plugin/Field/FieldWidget/RgbColorWidget.php:78]: Where is this function called?

File: web/modules/custom/my_field/config/schema/my_field.schema.yml (hunk 1 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/config/schema/my_field.schema.yml]: Removed

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BackgroundColorFormatter.php (hunk 2 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BackgroundColorFormatter.php]: Removed

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BodyValueFormatter.php (hunk 3 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BodyValueFormatter.php]: This is not something of the task, I was just exploring how to fetch a field's value and apply the formatted colour to that.
However updated it with the if clause.

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/TextColorFormatter.php (hunk 4 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/TextColorFormatter.php]: Didn't get this. Please Elaborate.

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/TextColorFormatter.php (hunk 5 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/TextColorFormatter.php]: Updated

File: web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php (hunk 6 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php:28]: Didn't get this, please elaborate.

Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php]: Rectified  the typing mistake

File: web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php (hunk 8 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php]: updated

File: web/modules/custom/my_field/src/Plugin/Field/FieldWidget/ColorPickerWidget.php (hunk 9 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldWidget/ColorPickerWidget.php:26]: Screenshot from 2023-08-09 10-58-44

File: web/modules/custom/my_field/src/Plugin/Field/FieldWidget/HexcodeWidget.php (hunk 10 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldWidget/HexcodeWidget.php:28]: Screenshot from 2023-08-09 11-00-33

File: web/modules/custom/my_field/src/Plugin/Field/FieldWidget/RgbColorWidget.php (hunk 11 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldWidget/RgbColorWidget.php:78]: The "massageFormValues" method is called during the process of preparing field values for storage in the database. The  method is typically called by Drupal's core field system automatically during the process of saving form data to the database.

File: web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php (hunk 7 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldType/FieldColor.php:30]: "columns" contains the values that the field will store. In this case I want to add a column in the database named as hexcode, which will be of type varchar and length 10.

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BodyValueFormatter.php (hunk 3 above)
Comment (vivekpan27) [reply web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BodyValueFormatter.php]: Even if it was not something from task, code should  be proper no?

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/TextColorFormatter.php (hunk 4 above)
Comment (vivekpan27) [reply web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/TextColorFormatter.php]: ID should be text_color_formatter.

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BodyValueFormatter.php [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  public function viewElements(FieldItemListInterface $items, $langcode) {
+
+    $elements = [];
+    foreach ($items as $delta => $item) {
+      // Storing the color.
+      $color = $item->hexcode;
+      // Storing that particuar entity.
+      $entity = $item->getEntity();
+      // Fetching that entity's body value.
+      if($entity->body->value)
```
Comment (vivekpan27) [inline web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BodyValueFormatter.php]: There should always be braces surrounding if statment even if it is a single line inside if.

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BodyValueFormatter.php (hunk 3 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BodyValueFormatter.php]: Yes sir.

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/TextColorFormatter.php (hunk 4 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/TextColorFormatter.php]: Updated

File: web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BodyValueFormatter.php (hunk 12 above)
Comment (Arijit-Acharya-innoraft) [reply web/modules/custom/my_field/src/Plugin/Field/FieldFormatter/BodyValueFormatter.php]: ok.
Updated

**Pull request merged**